File: microservices/recommendation/service/src/main/java/fr/edjaz/microservices/core/recommendation/services/RecommendationServiceImpl.java
package fr.edjaz.microservices.core.recommendation.services;

import fr.edjaz.api.core.recommendation.Recommendation;
import fr.edjaz.api.core.recommendation.RecommendationService;
import fr.edjaz.microservices.core.recommendation.persistence.RecommendationEntity;
import fr.edjaz.microservices.core.recommendation.persistence.RecommendationRepository;
import fr.edjaz.util.exceptions.InvalidInputException;
import fr.edjaz.util.http.ServiceUtil;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Slf4j
@RestController
public class RecommendationServiceImpl implements RecommendationService {

    private final RecommendationRepository repository;

    private final RecommendationMapper mapper;

    private final ServiceUtil serviceUtil;

    @Autowired
    public RecommendationServiceImpl(RecommendationRepository repository, RecommendationMapper mapper, ServiceUtil serviceUtil) {
        this.repository = repository;
        this.mapper = mapper;
        this.serviceUtil = serviceUtil;
    }

    @Override
    public Recommendation createRecommendation(Recommendation body) {

        if (body.getProductId() < 1) throw new InvalidInputException("Invalid productId: " + body.getProductId());

        RecommendationEntity entity = mapper.apiToEntity(body);
        Mono<Recommendation> newEntity = repository.save(entity)
            .log()
            .onErrorMap(
                DuplicateKeyException.class,
                ex -> new InvalidInputException("Duplicate key, Product Id: " + body.getProductId() + ", Recommendation Id:" + body.getRecommendationId()))
            .map(e -> mapper.entityToApi(e));

        return newEntity.block();
    }

    @Override
    public Flux<Recommendation> getRecommendations(int productId) {

        if (productId < 1) throw new InvalidInputException("Invalid productId: " + productId);

        return repository.findByProductId(productId)
            .log()
            .map(e -> mapper.entityToApi(e))
            .map(e -> {e.setServiceAddress(serviceUtil.getServiceAddress()); return e;});
    }

    @Override
    public void deleteRecommendations(int productId) {

        if (productId < 1) throw new InvalidInputException("Invalid productId: " + productId);

        LOG.debug("deleteRecommendations: tries to delete recommendations for the product with productId: {}", productId);
        repository.deleteAll(repository.findByProductId(productId)).block();
    }
}

File: microservices/recommendation/service/src/main/kotlin/fr/edjaz/microservices/core/recommendation/services/RecommendationServiceImpl.kt
package fr.edjaz.microservices.core.recommendation.services

import fr.edjaz.api.core.recommendation.Recommendation
import fr.edjaz.api.core.recommendation.RecommendationService
import fr.edjaz.microservices.core.recommendation.persistence.RecommendationEntity
import fr.edjaz.microservices.core.recommendation.persistence.RecommendationRepository
import fr.edjaz.util.exceptions.InvalidInputException
import fr.edjaz.util.http.ServiceUtil
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.DuplicateKeyException
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux

@RestController
class RecommendationServiceImpl @Autowired constructor(
    private val repository: RecommendationRepository,
    private val mapper: RecommendationMapper,
    private val serviceUtil: ServiceUtil
) : RecommendationService {
    companion object {
        @Suppress("JAVA_CLASS_ON_COMPANION")
        @JvmStatic
        private val logger = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

    override fun createRecommendation(body: Recommendation): Recommendation? {
        if (body.productId < 1) throw InvalidInputException("Invalid productId: " + body.productId)
        val entity = mapper.apiToEntity(body)
        val newEntity = repository.save(entity)
            .log()
            .onErrorMap(
                DuplicateKeyException::class.java
            ) { InvalidInputException("Duplicate key, Product Id: " + body.productId + ", Recommendation Id:" + body.recommendationId) }
            .map { e: RecommendationEntity -> mapper.entityToApi(e) }
        return newEntity.block()
    }

    override fun getRecommendations(productId: Int): Flux<Recommendation?>? {
        if (productId < 1) throw InvalidInputException("Invalid productId: $productId")
        return repository.findByProductId(productId)
            .log()
            .map { e: RecommendationEntity -> mapper.entityToApi(e) }
            .map { e: Recommendation? ->
                e!!.serviceAddress = serviceUtil.serviceAddress
                e
            }
    }

    override fun deleteRecommendations(productId: Int) {
        if (productId < 1) throw InvalidInputException("Invalid productId: $productId")
        logger.debug(
            "deleteRecommendations: tries to delete recommendations for the product with productId: {}",
            productId
        )
        repository.deleteAll(repository.findByProductId(productId)).block()
    }
}
